### PR TITLE
adding auth0 authentication with callback, login, logout views

### PR DESCRIPTION
**Pull request by phoenixAja**
### Description

adding in auth0 config

#### Issue
[ch64740](https://app.clubhouse.io/genepi/stories/space/64740)

### Test plan
hit login/logout endpoints try to hit a protected view when authenticated/ not authenticated, also deployed to eb-env `covidr-staging`

### CODE REVIEW

File: src/py/covidr/app.py [hunk 1]
```diff
@@ -1,25 +1,104 @@
 import os
 from pathlib import Path
+from functools import wraps
+
+from dotenv import find_dotenv, load_dotenv
+from flask import Flask, send_from_directory, redirect, session, url_for
+from authlib.integrations.flask_client import OAuth
+from six.moves.urllib.parse import urlencode
```
Comment (ttung) [inline src/py/covidr/app.py]: We don't support python 2, so we can just import urllib.parse's urlencode.

File: src/py/requirements.txt [hunk 2]
```diff
@@ -1,8 +1,12 @@
...
-itsdangerous==1.1.0
 Jinja2==2.11.2
 MarkupSafe==1.1.1
 SQLAlchemy==1.3.22
 Werkzeug==1.0.1
+click==7.1.2
+itsdangerous==1.1.0
+python-dotenv==0.15.0
+requests==2.24.0
+six==1.15.0
```
Comment (ttung) [inline src/py/requirements.txt]: You should edit `src/py/requirements.txt.in`, add in python-dotenv and Authlib, and run `make -j pin-requirements`

**Review by ttung — CHANGES_REQUESTED** [review]

File: src/py/requirements.txt (hunk 2 above)
Comment (phoenixAja) [reply src/py/requirements.txt]: oh ok so in the future we add new requirements to requirements.txt.ini and make -j pin-requirements will add them to requirements.txt? Can I ask what the benefit is of structuring it this way?

File: src/py/covidr/app.py (hunk 1 above)
Comment (phoenixAja) [reply src/py/covidr/app.py]: good point

File: src/py/requirements.txt (hunk 2 above)
Comment (ttung) [reply src/py/requirements.txt]: Yeah, we are tracking the requirements we actually need.  The transitive dependencies are pinned but kept separate.

File: src/py/covidr/app.py [hunk 3]
```diff
@@ -1,25 +1,104 @@
 import os
 from pathlib import Path
+from functools import wraps
+
+from dotenv import find_dotenv, load_dotenv
+from flask import Flask, send_from_directory, redirect, session, url_for
+from authlib.integrations.flask_client import OAuth
+from urllib.parse import urlencode
```
Comment (ttung) [inline src/py/covidr/app.py]:
```suggestion
import os
from functools import wraps
from pathlib import Path
from urllib.parse import urlencode

from dotenv import find_dotenv, load_dotenv
from flask import Flask, send_from_directory, redirect, session, url_for
from authlib.integrations.flask_client import OAuth
```

File: src/py/covidr/app.py [hunk 4]
```diff
@@ -1,25 +1,104 @@
...
 
 from covidr.config import DevelopmentConfig
-from flask import Flask, send_from_directory
+
+
+ENV_FILE = find_dotenv()
+if ENV_FILE:
+    load_dotenv(ENV_FILE)
```
Comment (ttung) [inline src/py/covidr/app.py]: Could you add documentation for what needs to be in the `.env` file and where it's supposed to be located?

File: src/py/covidr/app.py [hunk 5]
```diff
@@ -1,25 +1,104 @@
...
+
+AUTH0_CALLBACK_URL = os.environ.get("AUTH0_CALLBACK_URL")
+AUTH0_CLIENT_ID = os.environ.get("AUTH0_CLIENT_ID")
+AUTH0_CLIENT_SECRET = os.environ.get("AUTH0_CLIENT_SECRET")
+AUTH0_DOMAIN = os.environ.get("AUTH0_DOMAIN")
+AUTH0_BASE_URL = "https://" + AUTH0_DOMAIN
+SECRET_KEY = os.environ.get("SECRET_KEY")
```
Comment (ttung) [inline src/py/covidr/app.py]: I suspect many of these could live in the Config object.

**Review by ttung — CHANGES_REQUESTED** [review]

File: src/py/covidr/app.py (hunk 4 above)
Comment (phoenixAja) [reply src/py/covidr/app.py]: yes that is a very good idea

Comment (phoenixAja) [reply src/py/covidr/app.py]: do you think this would be better in the README or in the backend docs?

Comment (ttung) [reply src/py/covidr/app.py]: I imagine that README is a top-level directory plus anything that caters to a broad audience.  So in this case, the backend docs makes more sense.

**Comment by phoenixAja** [comment]
hmm for some reason this is breaking when deploying to ebs but was working fine when running locally, in the process of debugging

File: src/py/covidr/app.py (hunk 5 above)
Comment (lvreynoso) [reply src/py/covidr/app.py]: +1

**Comment by phoenixAja** [comment]
> hmm for some reason this is breaking when deploying to ebs but was working fine when running locally, in the process of debugging

fixed now, thanks @ttung for the help!

File: src/py/covidr/app.py [hunk 6]
```diff
@@ -1,25 +1,91 @@
...
+# use this to wrap protected views
+def requires_auth(f):
+    @wraps(f)
+    def decorated(*args, **kwargs):
+        if "profile" not in session:
+            # Redirect to Login page
+            return redirect("/login")
+        return f(*args, **kwargs)
+
+    return decorated
```
Comment (ttung) [inline src/py/covidr/app.py:46]: Not necessary to do in this PR, but putting this here could potentially cause problems with recursive imports if we break up `app.py`.

File: docs/backend/README.md [hunk 7]
```diff
@@ -1,3 +1,13 @@
+## Setup Auth0:
+
+you will need to have a specific `.env` file to read in `AUTH0` variables (place this file in `src/py/covidr`), ask project leads for this file, there will be different `.env` files for local development, staging, and production. 
+
+if deploying to a custom staging environment change this variable in the .env file:
+
+`AUTH0_CALLBACK_URL="http://<custom-staging-env-name>.us-west-2.elasticbeanstalk.com/callback"`
+
+if you are new to working on covidr ask a project lead to ask your eb staging env to the list of allowed callbacks in Auth0. 
+
```
Comment (ttung) [inline docs/backend/README.md:10]: I created a couple follow-on stories:

https://app.clubhouse.io/genepi/story/68840/fold-authconfig-into-main-config-object
https://app.clubhouse.io/genepi/story/68841/figure-out-how-to-set-env-variables-as-part-of-eb-deploy

**Review by ttung — APPROVED** [review]

**Pull request merged**